pkg/controller/pv-monitor-controller/node_watcher.go

```diff
@@ -0,0 +1,455 @@
+	// TODO: if node is ready, check if node was ever marked down, if yes, reset it
+	if watcher.isNodeReady(node) {
+		// The node status is ok, but if it was marked before, remove the mark
+		_, ok := watcher.nodeFirstBrokenMap[node.Name]
+		if ok {
+			delete(watcher.nodeFirstBrokenMap, node.Name)
+		}
+
+		// if the node was ever marked down, reset PVCs status on it
+		if watcher.nodeEverMarkedDown[node.Name] {
+			// TODO: reset PVCs status on the node
+			err := watcher.cleanNodeFailureConditionForPVC(node)
+			if err == nil {
+				// when node recovers and send recovery event successfully, remove the node from the map
+				delete(watcher.nodeEverMarkedDown, node.Name)
+			} else {
+				klog.Errorf("clean node failure message error: %+v", err)
+			}
+		}
+
+		return
+	}
+
+	// TODO: need to revisit this later
+	if watcher.isNodeBroken(node) {
+		klog.Infof("node: %s is broken", node.Name)
+		// mark all PVCs/Pods on this node
+		// try several times again
+		var err error
+		for i := 0; i < UpdatePVRetryCount; i++ {
+			err = watcher.markPVCsAndPodsOnUnhealthyNode(node)
+			if err != nil {
+				time.Sleep(UpdatePVInterval)
+				continue
+			}
+			break
+		}
+		if err != nil {
+			klog.Infof("mark PVCs on not ready node failed, re-enqueue")
+			// if error happened, re-enqueue
+			watcher.enqueueWork(node)
+			return
+		}
+
+		// node is broken and PVCs on it are marked, remove it from map
+		_, ok := watcher.nodeFirstBrokenMap[node.Name]
+		if ok {
+			delete(watcher.nodeFirstBrokenMap, node.Name)
+		}
+
+		watcher.nodeEverMarkedDown[node.Name] = true
+	}
+
+}
+
+func (watcher *NodeWatcher) isNodeReady(node *v1.Node) bool {
+	for _, condition := range node.Status.Conditions {
+		if condition.Type == v1.NodeReady && condition.Status == v1.ConditionTrue {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (watcher *NodeWatcher) isNodeBroken(node *v1.Node) bool {
+	if node.Status.Phase == v1.NodeTerminated {
+		return true
+	}
+	objName := node.Name
+	for _, condition := range node.Status.Conditions {
+		if condition.Type == v1.NodeReady && condition.Status != v1.ConditionTrue {
+			now := time.Now()
+			firstMarkTime, ok := watcher.nodeFirstBrokenMap[objName]
+			if ok {
+				timeInterval := now.Sub(firstMarkTime)
+				if timeInterval.Seconds() > DefaultNodeNotReadyTimeDuration.Seconds() {
+					return true
+				} else {
+					klog.V(6).Infof("node:%s is not ready, but less than 5 minutes, re-enqueue", node.Name)
+					// NotReady status lasts less than 5 minutes
+					// re-enqueue
+					watcher.enqueueWork(node)
+					return false
+				}
+			} else {
+				// first time to mark the node NotReady
+				watcher.nodeFirstBrokenMap[objName] = now
+				watcher.enqueueWork(node)
+				return false
+			}
+		}
+	}
+
+	return false
+}
+
+func (watcher *NodeWatcher) deleteNode(key string, node *v1.Node) {
+	klog.Infof("node:%s is deleted, so mark the local PVs on it", node.Name)
+
+	// mark all local PVs on this node
```

Comment:
Is NodeWatcher only for Local PVs?  Will this be useful for other volumes?  Either way, we should add comments to the NodeWatcher enable option about this.

Reply:
done